src/main/java/org/folio/fqm/service/EntityTypeValidationService.java

```diff
@@ -114,47 +125,125 @@ public void validateEntityType(UUID entityTypeId, EntityType entityType, List<UU
+    List<UUID> validTargetIds
+  ) {
+    UUID targetId = source.getTargetId();
     if (targetId == null) {
-      throw new InvalidEntityTypeDefinitionException("Source entity type ID cannot be null for entity-type sources", entityType);
+      throw new InvalidEntityTypeDefinitionException(
+        "Source %s's target ID cannot be null".formatted(source.getAlias()),
+        entityType
+      );
     }
     if (validTargetIds == null) {
-      if (entityTypeRepository.getEntityTypeDefinition(targetId, folioExecutionContext.getTenantId())
-        .isEmpty()) {
+      if (entityTypeRepository.getEntityTypeDefinition(targetId, folioExecutionContext.getTenantId()).isEmpty()) {
         throw new InvalidEntityTypeDefinitionException(
-            "Source with target ID " + targetId + " does not correspond to a valid entity type", entityType);
+          "Source %s's target ID %s refers to an unknown entity type".formatted(source.getAlias(), targetId),
+          entityType
+        );
       }
     } else if (!validTargetIds.contains(UUID.fromString(targetId.toString()))) {
       throw new InvalidEntityTypeDefinitionException(
-          "Source with target ID " + targetId + " does not correspond to a valid entity type", entityType);
+        "Source %s's target ID %s refers to an unknown entity type".formatted(source.getAlias(), targetId),
+        entityType
+      );
+    }
+
+    validateEntityTypeSourceJoinProperties(entityType, source);
+  }
+
+  private static void validateEntityTypeSourceJoinProperties(EntityType entityType, EntityTypeSourceEntityType source) {
+    List<Boolean> possessedRequiredJoinProperties = List.of(
+      source.getTargetField() != null,
+      source.getSourceField() != null
+    );
+    long numRequiredJoinProperties = possessedRequiredJoinProperties.size();
+    long numPossessedRequiredJoinProperties = possessedRequiredJoinProperties
+      .stream()
+      .filter(Boolean.TRUE::equals)
+      .count();
+    boolean hasAllRequiredJoinProperties = numPossessedRequiredJoinProperties == numRequiredJoinProperties;
+    boolean hasNoRequiredJoinProperties = numPossessedRequiredJoinProperties == 0;
+
+    boolean hasOptionalJoinProperties = source.getOverrideJoinDirection() != null; // optional
+
+    // we only have some, not all
+    if (!hasAllRequiredJoinProperties && !hasNoRequiredJoinProperties) {
+      throw new InvalidEntityTypeDefinitionException(
+        "Source %s must contain both targetField and sourceField or neither".formatted(source.getAlias()),
+        entityType
+      );
+    }
+
+    // case with optional + some properties is already covered above
+    if (hasOptionalJoinProperties && hasNoRequiredJoinProperties) {
+      throw new InvalidEntityTypeDefinitionException(
+        "Source %s may only contain overrideJoinDirection if targetField and sourceField are also defined".formatted(
+            source.getAlias()
+          ),
+        entityType
+      );
```

Comment:
This is probably overkill, but it is versatile for if/when we add more join properties in the future